Azure Container Instances - closes #511 (#512)
bob-crutchley/courseware: topics/azure/modules/container-instances/README.md

# Container Instances

<!--TOC_START-->
## Contents
- [Overview](#overview)
- [Why ACI?](#why-aci)
- [Using ACI](#using-aci)
- [Using exec](#using-exec)
- [Logs](#logs)
- [The Portal](#the-portal)
- [Cleaning Up](#cleaning-up)
- [Task](#task)

<!--TOC_END-->
## Overview

> Azure Container Instances (ACI) allow you to focus on designing and building your applications instead of managing the infrastructure that runs them.

Reviewer: ```suggestion
**A**zure **C**ontainer **I**nstances (**ACI**) allow you to focus on designing and building your applications instead of managing the infrastructure that runs them.
```

Author: thanksss

> ACI are especially useful when you want to run a container without orchestration, which usually means you don't need to scale the application or utilise load balancing.

Reviewer: ```suggestion
*ACI* are especially useful when you want to run a container without orchestration, which usually means you don't need to scale the application or utilize load balancing.
```

## Why ACI?

The advantage of ACI over utlising your own containers is the managed infrastructure.

Typically, if you wanted to run an application in a container, you would need to either:

1. Use your local machine
2. Spin up a virtual machine (VM), install Docker and run the container

> Both of these ways mean you have to manage the infrastructure yourself. ACI takes that away, with Miocrosoft managing and configuring the VM for you. You don't even see it in this entire process!

Reviewer: ```suggestion
Both of these ways mean you have to manage the infrastructure yourself. 

ACI takes that away, with Miocrosoft managing and configuring the VM for you. 

You don't even see it in this entire process!
```

All you need is a Docker image: this can either be pulled from your own DockerHub or another public registry. 

> Azure actually has a container registry (Azure Container Registry) you can make use of when working with ACI if you'd prefer to use this over DockerHub.

Reviewer: ```suggestion
Azure actually has a container registry called **Azure Container Registry** you can make use of when working with ACI if you'd prefer to use this over DockerHub.
```

# Using ACI

In this module, we are going to use the az CLI to work with ACI. 

We will be using a public Docker image that has the game 'Super Mario' on it.

The first thing we need to do is create a resource group:

```powershell

> az group create -n mariogroup -l uksouth

Reviewer: explain what the flags mean for the command

```

Next, we need to create the container and give it the correct image to use:

```powershell

> az container create -g mariogroup -l uksouth -n mario --image jordangrindrod/mario --ip-address public --ports 8080

Reviewer: explain what the flags mean for the command

```

Here, we have ensured that the VM Azure spins up will have a Public IP address and that the application will be running on port 8080.

Once this has completed, you can run the following the see your running container and the ip address of it:

```powershell

> az container list -o table

Reviewer: explain what the flags mean for the command

```

Then, you can navigate to the browser and paste the ip address in, followed by `:8080`.

Have a couple of minutes playing Mario, then come back to the lesson...

## Using exec

Microsoft have leveraged Docker's `exec` command, so that you can interact with your container directly.

For example, we could have a look at our container's filesystem with the following command:

> ```powershell

Reviewer: explain what the flags mean for the command

az container exec -g mariogroup -n mario --exec-command ls
```

## Logs

The following command allows you to see the logs of this container instance, which can be incredibly useful for debugging should something go wrong:

```powershell

> az container logs -g mariogroup -n mario

Reviewer: explain what the flags mean for the command

```

## The Portal

If you navigate to Container Instances in the Azure Portal, you will see your `mario` container up and running.

Metrics are automatically linked to each container instances you create, allowing you to see data such as CPU usage. 

You can also view IAM, the logs for each instance, add or remove tags, and much more.

## Cleaning Up

To remove everything you have done in this lesson, simply run the following command:

```powershell

> az group delete -n mariogroup -y

Reviewer: explain what the flags mean for the command

```

## Task

Now it is time for you to try this for yourself:

> 1. Create a new Resource Group

Reviewer: This is where usually you would guide people step by step, so in this case it would be you to give them commands and tell them what to replace to make it work

additionally I don't believe you specified how to make the commands work, like how to install the AZ cli locally or how to use the one in the AZ portal, would be good if you included that

Author: I kinda already did the step by step in the lesson - they just need to replicate it for themselves now

Also I was going to do the AZ CLI install and that but they should do that before they do this lesson - there will be modules in week 1 for it. So i thought that would be considered a pre-requisite for this module, instead of keep replicating information?
